students/ahwall/Lesson06Homework/Lesson06HW_Git/MailRoomPt3.py

```diff
@@ -0,0 +1,167 @@
+                try:
+                    DonationVal = input("\nAdding Donor to list,\n\nHow much was \
+their donate?  ")
+                #  place exception to make sure float is given
+                    DonationVal = float(DonationVal)
+                    vallist = []
+                    vallist.append(DonationVal)
+                except ValueError:
+                    return ("\nYou did not enter an numberic value, Please try \
+again\n")
+                    # thank_you()
+                else:
+                    Don_Val = ()
+                    Don_Val = (ty_letter[:], vallist[:])
+                    donors.append(Don_Val)
+                    print()
+                    return (ty_letter, "and their donation have been added to \
+the Donor Report...Creating 'Thank you email'")
+                    i = 1
+                    One_letter(ty_letter, DonationVal)
+                    mainloop()
+                break  # Works
+
+
+def One_letter(donor, donation):
+    print()
+    return "Dear {};\n\n\tThank you for your generous donation of ${:.2f}.\n\n \
+\tSincerely,\n\n \tThe Python Project\n".format(donor, donation)  # Works
+
+
+def All_letters(donordict):
+    donorvallist = list(donordict.values())
+    for i in range(len(donorvallist[0])):
+        Donor, Donation = donordict['Donor'][i], donordict['Donation'][i]
+        Donation = int(Donation)
+        # could not figure out a way to unpack the dictionary (like Mr. Barker
+        # suggested in the question) with multiple values per key, since you
+        # can not use [i] because it needs to be a int not a string
+        lettertemplate = "Dear {},\n\n\
+        Thank you for your very kind donation of ${:,.02f}. \
+        \n\n\t\t Sincerely,\n\t\t  -The Team".format(Donor, Donation)
+        filename = open("{}.txt".format(Donor), 'w')
+        filename.write(lettertemplate)
+        filename.close()
+    print("Creating 'Thank you' letters to all Donors")
+    #  mainloop()
+    #  could not get it to pass test_All_letters with mainloop() still active
+
+
+#  Could not get it to pass Test_report Unit test using comprehension list
+def report(donors):
+    Repheader = ('Donor Name', 'Total Given', 'Num Gifts', 'Average Gifts')
+    headerspace = '{:20}| {:^14}| {:^10}| {:^14}'
+    ReportCats = headerspace.format(*Repheader)
+    print('\n', ReportCats)
+    print('-' * 63)
+    # for loop to print report
+    don_report = []
+    for item in donors:
+        don_report = ('{:20}| ${:^13,.02f}| {:^10}| ${:^14,.02f}'.
+        format(item[0], sum(item[1]), len(item[1]), (sum(item[1]) /
+        len(item[1]))))
+        print(don_report)
+    return don_report
+    #don_report = [print('{:20}| ${:^13,.02f}| {:^10}| ${:^14,.02f}'.format(item[0],
+    #sum(item[1]), len(item[1]), (sum(item[1]) / len(item[1]))))
+    #for item in donors]
+    #  return don_report
+    #  mainloop()
+    # could not get it to pass test_All_letters with mainloop() still active
```

Comment:
Oh yeah, and calling to mainloop() builds up the stack, so you don't really want to do that, so yay for test-driven-development